client/checkout/classic/upe.js

```diff
@@ -453,14 +453,18 @@ jQuery( function ( $ ) {
 
 		try {
 			// Update payment intent with level3 data, customer and maybe setup for future use.
-			await api.updateIntent(
+			const updateResponse = await api.updateIntent(
 				paymentIntentId,
 				orderId,
 				savePaymentMethod,
 				$( '#wcpay_selected_upe_payment_type' ).val(),
 				$( '#wcpay_payment_country' ).val()
 			);
 
+			if ( updateResponse.data ) {
+				await api.handlePreviousOrderPaid( updateResponse.data );
```

Comment:
> No need for await here, handlePreviousOrderPaid is not asynchronous.

I used `await` to avoid the risk you mentioned. That is, when the redirection has not finished yet, the code continues to confirm payment, and yell an error like the following.

![Markup on 2022-12-23 at 17:57:47](https://user-images.githubusercontent.com/10045087/209324807-1aa267a2-dfed-4d17-8565-9f8130e91946.png)

But yeah, it does not sound the correct way to handle this. 

> However, it is also worth returning something from handlePreviousOrderPaid, checking it here and bailing if an order has been handled.

This is the right way. Addressed here a3aa50bf648abbf99c38e99e4f424f2a8468f0a5

Reply:
Marking this conversation as resolved.